packages/compass-connections/src/connection-info-provider.tsx

```diff
@@ -1,48 +1,29 @@
-import React, { createContext, useContext, useState } from 'react';
-import type { ConnectionInfo } from '@mongodb-js/connection-info';
+import React, { createContext, useContext, useRef } from 'react';
+import { type ConnectionInfo } from '@mongodb-js/connection-info';
 import {
   createServiceLocator,
   createServiceProvider,
 } from 'hadron-app-registry';
 import {
-  ConnectionsStoreContext,
   useConnectionForId,
+  useConnectionInfoForId,
   useConnectionInfoRefForId,
 } from './stores/store-context';
-import type { ConnectionId } from './stores/connections-store-redux';
+import type {
+  ConnectionId,
+  ConnectionState,
+} from './stores/connections-store-redux';
 
 export type { ConnectionInfo };
 
-export type ConnectionInfoAccess = {
-  getCurrentConnectionInfo(): ConnectionInfo;
-};
-
-const ConnectionInfoContext = createContext<ConnectionInfo | null>(null);
-
 const ConnectionIdContext = createContext<ConnectionId | null>(null);
 
 /**
- * @deprecated define connection for your test separately
+ * @internal never to be used outside of the test environment or re-exported
+ * from this package
  */
-export const TEST_CONNECTION_INFO: ConnectionInfo = {
-  id: 'TEST',
-  connectionOptions: {
-    connectionString: 'mongodb://localhost:27020',
-  },
-};
-
-export function useConnectionInfo() {
-  const connectionInfo = useContext(ConnectionInfoContext);
-  if (!connectionInfo) {
-    if (process.env.NODE_ENV !== 'test') {
-      throw new Error(
-        'Could not find the current ConnectionInfo. Did you forget to setup the ConnectionInfoContext?'
-      );
-    }
-    return TEST_CONNECTION_INFO;
-  }
-  return connectionInfo;
-}
+export const TestEnvCurrentConnectionContext =
+  createContext<ConnectionState | null>(null);
```

Comment:
I like to avoid test environments making their way into the running code, I'm trying to understand the reason why we want this instead of having the a connection in the state of the connections store context. Why do we have this? Is it so that we avoid any interactions with the connection and any connection ids?

Not a blocker.

Reply:
Due to the nature of hooks that introduce a very tight coupling within the components and the nature of this particular state that is required almost everywhere in the application we are left with more or less the following options:

- Stop writing unit and functional tests completely, only write e2e tests that make sure that the app is in the correct state that allows components to render following the application logic to the letter
- Introduce explicit handling for those cases in the application business logic as if it's part of the actual user flows, account for that, document and test is as if it's a real app behavior, consider the implications for real user-facing features we implement going forward
- Introduce explicit unit / functional test environment switches, providers, and various other helpers so that we can continue writing unit / functional tests for the user facing features instead of testing unrelated pieces of the application logic over and over again

All options come with their own set of benefits and trade-offs. From my perspective this option we're going for here enables us to write solid unit tests while removing the need to manually run the whole application every time you need to test an isolated piece of the code that doesn't really care about having a real connection established. This comes with the downside of having extra code in the application that is accounting for the test env outside of the main application business logic path, that's not great, but we are in control of how it's being used, this particular code is not even exported from the package for example, so using it would indeed require a pretty intentional breaking of a lot of rules we have in this codebase.

I would probably not recommend going for the first option at all due to the complexity of writing and maintaining a robust set of stable e2e tests for every small piece of code we have in the app. We can consider the second option, but it has product implications and potentially introduces a risk of these APIs being misused in the real production code as opposed to scoping this concern only to test environment

Reply:
To clarify even more maybe, I do think that the second option is sort of viable and can be mixed in with the third one, you do need to weight in the implications it introduces for real app business logic every time though.

For example, if I ask "Is there any harm in a public interface to allow to syncronously set initial connections from the storage without waiting for connections to load when component mounts?", I think the answer here is "there is very litte to no harm in that, this info is only used to list connections, not operate on them", so that's what we're using to allow for the default `render` function to be syncronous and not wait for the list of connections initial load.

At the same time if I ask myself "Is there any harm in providing an interface for the connection info provider to allow to override the connection info that is being accessed by the id?", I think the answer is "yes, it seems scary as it might allow to completely override the connection being used by a feature in the app without being able to control or spot this easily", so a dedicated provider for test env is used here without exposing the interface.